### ISSUE TITLE
[Bug]: `useParameter` behaviour as you change between stories

### ISSUE THREAD

**Issue by tmeasday** [storybookjs/storybook#22119]
### Describe the bug

- Currently the manager only knows the parameter of stories it has rendered before.
- When you select a story it emits a `STORY_PREPARED` event which reaches the manager after some delay.
- Many addons (such as `backgrounds`) cannot work, and must hide themselves if they don't see the right parameter.

The net effect of this is when you browse between stories, there is a "flash of hidden toolbars" as the addons hide themselves while they wait for the new story to prepare.

### To Reproduce

- Refresh the browser
- Browse around a sandbox
- Observe the backgrounds addon carefully.

### Additional context

To fix this, we might consider:

**Falling back to the component + project parameters** (we could make them available ala this unmerged PR: https://github.com/storybookjs/storybook/pull/22071)

This generally speaking would work well, but in some examples it might have similar issues of things flashing on and off as it loads. For example when you move between two stories with non-standard settings for the add (perhaps hiding it), the addon would briefly appear.

**Continuing to return the previous story's value until the new story is prepared**.

This might be complex to implement, but could be the best solution.

**Comment by tmeasday** [comment]
Hey @raphaelokon, yes! Sorry I didn't link them together yet. This is basically a follow-on issue from that discussion that we'll think about for 7.1. Because we are in post-release mode for 7.0 we are trying to make the change there as small as possible, I hope that makes sense!

**Comment by probablyup** [comment]
Following up from the other issue:

The dark mode addon uses parameters for configuration storage, and I imagine many other addons do as well. Specifically, there is a darkClass and lightClass that can be user-supplied to match the selector used for scoping of light/dark mode styles in userland.

If docs stories do not have access to the parameters provided in preview.js, then there will be an unavoidable white FOUC for a user in dark mode as the iframe initializes that is jarring and did not previously happen under Storybook v6 (at least with storyStoreV7 off, which anyone who requires dynamic stories is likely to have configured.)

**Comment by tmeasday** [comment]
> If docs stories do not have access to the parameters provided in preview.js, then there will be an unavoidable white FOUC for a user in dark mode as the iframe initializes that is jarring and did not previously happen under Storybook v6 (at least with storyStoreV7 off, which anyone who requires dynamic stories is likely to have configured.)

I don't think this is correct. You'd still be waiting for the initial `SET_STORIES` (v6) event to get parameters, which conceivably would take even longer than the initial `STORY_CONFIGURED`/`DOCS_CONFIGURED` (v7) event. A FOUC on startup is unavoidable if the addon relies on data supplied by the preview.

**Comment by tmeasday** [comment]
@probablyup there is a FOUC, it's just that the addon uses local storage to avoid it after the first render.

![Screen Recording 2023-04-19 at 10 55 33 am](https://user-images.githubusercontent.com/132554/232961431-87a05965-cef0-4f72-9f1d-c76f48a05561.gif)

I just tried `6.5.x` and `7.1.0-alpha.6` and AFAICT the behaviour is unchanged now 🚀 (let me know if I missed something). Try these two branches:

 - `6.5.x`: https://github.com/tmeasday/test-dark-mode/tree/main
 - `7.1.0-alpha.6`: https://github.com/tmeasday/test-dark-mode/tree/dark-mode-7

**Comment by probablyup** [comment]
> Tags are intentionally simple and simply a list of strings, so might not hit every use case, but let's see how we go!

Unless you can provide a config object in tags, I don’t see how that solves the problem…

Would it be possible to just introduce some blessed API that does what is needed: global addon config storage?

**Comment by tmeasday** [comment]
> Would it be possible to just introduce some blessed API that does what is needed: basic k-v storage available before stories are rendered that can be bootstrapped from preview.js

Well project parameters could be that @probablyup, right? We don't currently send them to the manager on their own (I did briefly in https://github.com/storybookjs/storybook/pull/22071), but if that's the blessed way to share project-level configuration between preview + manager, then we could add the `PROJECT_PREPARED` event back. Do you have thoughts on this @shilman?

The use case we were discussing (and really what this issue is about) is around configuring things *per-story* differently. For that, we have:
 - parameters, which only are available when the story is loaded, but can contain more or less anything.
 - tags, which are available for all stories, immediately, but are limited to a list of strings.

We could add another concept but so far it seems like tags hit a lot of the "per story configuration I need to know about all the time" pretty well.

**Comment by probablyup** [comment]
Ultimately what I’m looking to achieve is a way for an addon to have config that’s available immediately from anywhere in storybook, regardless of docs or component story type.

**Comment by ndelangen** [comment]
> a way for an addon to have config

Would this be **static** config? Or does this config change over time?

**Comment by probablyup** [comment]
> Would this be static config? Or does this config change over time?

for storybook-dark-mode the config doesn’t really change between pages, so I’d say it’s static. As soon as storybook initializes, the config is actually copied into localStorage to make things faster for subsequent visits

**Comment by ndelangen** [comment]
You could experiment with using the `env` feature.

This is a setting for webpack define plugin:
https://github.com/storybookjs/storybook/blob/d760a5a4c01a21441b0a22c5cefd71c1ae4899da/code/lib/builder-webpack5/src/preview/iframe-webpack.config.ts#L276
but also vite:
https://github.com/storybookjs/storybook/blob/a0d5ecf8ec0d02c9ac4e38720107ecc9ec0689f6/code/lib/builder-vite/src/envs.ts#L45-L54
and the manager too:
https://github.com/storybookjs/storybook/blob/4a37372f649e85e7a0c35b0493da016dbb5dee17/code/lib/builder-manager/src/index.ts#L97-L102

So if the data is stringifable (which it likely is, considering you're also putting it in local storage), that might work.

in your `main.ts` file, add a `env` entry, that's a async function.

```ts
export default {
  // .. 
  env: async (config) => ({
    ...config,
    SOME_UNIQUE_KEY: JSON.stringify({ your: 'value'}),
  })
};

I haven't tested this, I think it might work.

If you config object is very large this could have negative performance effects.

**Comment by shilman** [comment]
@tmeasday in DM you proposed a solution where the user can also export `parameters` from `.storybook/manager.js` to make sure they are instantly available in the manager.

The behavior would then be:
- manager.js provides parameter defaults which are available on load
- those parameters get overwritten by the preview parameters when stories are loaded

This means that you'll get a FOUC if you don't specify manager.js parameters OR if they differ from the story parameters. This feels like a really reasonable solution to me, though it does add some weird edge cases, e.g. will there be manager-only parameters if the user adds manager.js parameters that they don't add to the preview? do we still need localStorage? etc.

what does everybody else think?

**Comment by tmeasday** [comment]
So to be clear the solution @shilman is talking about is to define the them in a file that is imported both in `manager.js` and `preview.js`:

```ts
// preview.js
import { themes } from './themes.js';

export default {
  parameters: {
    darkMode: themes
  }
};
```

The main downside I can see with this is it's a little harder for the user to configure. However it's not *that* hard and probably adding a new concept to our API to deal with this problem is hard to swallow.

I think the main alternative suggestion is something automated that behind the scenes works like @ndelangen is suggesting, basically that you configure the addon in `main.js` and SB wires those options through to the manager+preview somehow. So something like:

```ts
// user main.js
export default {
  addons: [{ name: 'storybook-dark-mode', options: { dark: {...}, light: {...} } }],
}

// addon code, perhaps this API would work in both manager+preview contexts:
const { dark, light } = getAddonOptions('storybook-dark-mode');
```

**Comment by ndelangen** [comment]
@tmeasday that sounds like a great start for a API proposal tbh, should we create an RFC for that?

**Comment by tmeasday** [comment]
I guess so @ndelangen. I'm still unsure if it's useful enough to commit to it, but perhaps an RFC will help us tease that out.
